New Api Functions

gchoqueux opened the pull request:
**Api functions**
- Add `ApiGlobe.setZoomLevel()`
- Add` ApiGlobe.getZoomScale()`
- Add `ApiGlobe.setZoomScale(zoomScale)`
- Add `ApiGlobe.pan(pVector)`
- Partial Fix bug `ApiGlobe.removeImageryLayer()`

**Framework**
- Three js v80 to 82

**Core**
- SSE
 Remove small oblique weight in Screen space error compute [(see page 372 '3D engine design for virtual globes') ](http://www.virtualglobebook.com/3DEngineDesignForVirtualGlobesSection121.pdf)
 Oblique weight is wrong for tiles with big altitude

nosy-b commented:
- Error while calling `itowns.viewer.setZoomScale(zoomScale);` directly after loading the globe. If user click and move it then it works.
- `ApiGlobe.prototype.getZoomLevel` has bad documentation and sample code. It doesn"t take any parameters.

gchoqueux committed:
THREE v80 to v82

gchoqueux commented:
@nosy-b see new commits
make `npm install` to check `THREE v82`

nosy-b commented:
`ApiGlobe.SetCenter()` works but it doesn't keep camera upVector to current direction (North by default)

Other functions seems to work fine. Remember to add documentation for all functions (at least API ones)

gchoqueux commented:
> `ApiGlobe.SetCenter()` works but it doesn't keep camera upVector to current direction (North by default)
> use ApiGlobe.setCenterAdvanced() for the moment.
> Add a issue for this feature

The branch is updated

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -591,13 +607,111 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
+ * @param      {number}  pitch   Screen pitch, in millimeters ; 0.28 by default
+ * @return     {number}  The zoom scale.
+ */
+ApiGlobe.prototype.getZoomScale = function(pitch) {
```
What is the difference betweem zoom _level_ and zoom _scale_ ?

gchoqueux replied:
ZoomLevel, zoom camera based on level (tile matrix)
ZoomScale, zoom camera based on scale's map

peppsac replied:
So zoom level is based on WMTS' notion of levels right?

gchoqueux replied:
Zoom level is based level's tile mesh

peppsac replied:
I get the same result when using `itowns.viewer.setZoomScale(12)` or `itowns.viewer.setZoomScale(0.0000001)`.
But if I use `itowns.viewer.setZoomScale(0.00000001)` I get a black screen and then `setZoomScale` is broken (`rayon` is always `NaN`).

gchoqueux replied:
branch updated

peppsac commented on src/Scene/LayersConfiguration.js:
```diff
@@ -44,76 +44,76 @@ function defaultState(seq) {
 LayersConfiguration.prototype.removeColorLayer = function(id) {
     if (this.layersState[id]) {
-        this.colorLayers.filter(function(l) {
+        this.colorLayers = this.colorLayers.filter(function(l) {
```
The changes to this file should be done in a separate commit.

gchoqueux committed:
Fix bug ApiGlobe.removeImageryLayer()

gchoqueux commented:
branch updated

peppsac commented:
It's hard to test this PR: the API is not accessible using the GUI.
Maybe you could add a new `dat.gui` window with a preconfigured action for each API function?

gchoqueux commented:
> Maybe you could add a new dat.gui window with a preconfigured action for each API function?

Yes, in new PR.
If not, I use console's navigator to test API

gchoqueux commented:
branch updated and rebased

gchoqueux commented:
r?

peppsac commented:
Maybe we could start using label to mark PR that are ready for reviews?

peppsac commented:
Anyway I started looking at this one but I got lost investigating errors in `getPickingPosition` while testing:

``` js
window.itowns.viewer.setZoomLevel(12)
GlobeControls.js:595 Uncaught TypeError: Cannot read property 'distanceTo' of undefined(…)
```

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
  */
+ApiGlobe.prototype.setZoomLevel = function(zoom) {
+
+    zoom = Math.max(2,zoom);
```
Why do you use a hard-coded level?
You should compute `zoom` using `[getMinZoomLevel(), getMaxZoomLevel()]`

(and I just tested this: these 2 methods are broken :-( )

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -456,16 +469,17 @@ ApiGlobe.prototype.getHeading = function() {
+
+ApiGlobe.prototype.getRangeFromEllispoid = function() {
 
+    // TODO: error is distance is big with ellipsoid.intersection(ray) because d < 0
```
Typo in the method name `getRangeFromEllispoid` → `getRangeFromEllipsoid`
Also : no documentation? what is the difference between getRange and `getRangeFromEllipsoid`? why/when should I use one or the other?

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
+    let distance = rayon + range;
+    let alpha = camera.FOV / 180 * Math.PI * 0.5;
+    let phi = Math.PI - Math.asin(distance / rayon * Math.sin(alpha));
+    let projection;
```
Would you mind adding some comments about what is going on here?
It's not obvious, even if you're using sensible variable names.

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
+        projection = rayon * Math.sin(beta) * 2.0;
+        // factorisation -->> */
+        projection = 2.0 * rayon * Math.sin(phi + alpha);
+
```
Unneeded empty lines + dead code

gchoqueux replied:
I prefer keep the develop operation to understand : `projection = 2.0 * rayon * Math.sin(phi + alpha);`

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
+    let rayon = this.scene.currentControls().getCenter().length();
+    let alpha = camera.FOV / 180 * Math.PI * 0.5;
+    let distance;
+    let sinBeta = projection / (2 * rayon);
```
Same: this really needs some comments.

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -580,13 +594,117 @@ ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
+ApiGlobe.prototype.pan = function(pVector) {
+
+    this.scene.currentControls().pan(pVector.x,pVector.y);
+
```
Remove empty lines

gchoqueux replied:
ok

peppsac commented on src/Globe/Globe.js:
```diff
@@ -219,10 +220,18 @@ Globe.prototype.getZoomLevel = function( /*id*/ ) {
 
+
+Globe.prototype.computeDistanceForZoomLevel = function(zoom,camera) {
+
```
3 hard-coded values in a single line: 2 (min level), 18 (max level), 6.0 (sse).

Proposal  : 
- `2` becomes `this.tiles.minLevel`
- `18` becomes `this.tiles.maxLevel`
- `6.0` : you could add a `export const SSE_SUBDIVISION_THRESHOLD` in `NodeProcess.js` and replace all usage of `6.0` in itowns: `NodeProcess.js`, `QuadTree.js` and here.

(and empty blank lines)

peppsac commented on src/Renderer/Camera.js:
```diff
@@ -124,11 +124,13 @@ Camera.prototype.computeNodeSSE = function(node) {
     var boundingSphere = node.geometry.boundingSphere;
     var distance = Math.max(0.0, (this.camera3D.position.distanceTo(node.centerSphere) - boundingSphere.radius));
     // Added small oblique weight (distance is not enough, tile orientation is needed)
+    /*
```
You're switching back to the old SSE?
Documenting why (in commit message) would be interesting - and maybe change the commit's title btw: you're doing more than juste `AddNew Api Functions`.

gchoqueux replied:
the previous one was false calculation. I will return it later

gchoqueux replied:
I updated the branch with a new commit and a new comment in PR

peppsac replied:
You ignored the 2nd part of my comment :)

> Documenting why (in commit message) would be interesting - and maybe change the commit's title btw: you're doing more than juste AddNew Api Functions.

peppsac replied:
Oops, I forgot to read the commit messages.. nevermind :-)

gchoqueux replied:
and I add  link  in [ main comment PR #194](https://github.com/iTowns/itowns2/pull/194#issue-181982034)

peppsac commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -471,41 +469,39 @@ function GlobeControls(camera, domElement, engine) {
-        if (state === CONTROL_STATE.MOVE_GLOBE) {
+            movingGlobeTarget.copy(this.getCenter()).applyQuaternion(quatGlobe);
+            this.camera.position.copy(snapShotCamera.position).applyQuaternion(quatGlobe);
+            this.camera.up.copy(movingGlobeTarget.clone().normalize());
```
Again: adding a single line of comment explaining what you're doing here would be useful.
Examples: 

``` js
// Rotate globe then update camera to keep camera's center aligned
// Pan camera along globe's normal
```

peppsac replied:
The controls are still buggy (even after this PR: try clicking outside the globe), so adding comments explaining what the code is supposed to do will help us fix the problems.

gchoqueux replied:
For info on rework GlobeControl see #153 

> try clicking outside the globe

 Can you be more precise for this bug

peppsac commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -471,41 +469,39 @@ function GlobeControls(camera, domElement, engine) {
+            movingGlobeTarget.applyQuaternion(quaterPano);
+            this.camera.localToWorld(movingGlobeTarget);
+            this.camera.up.copy(movingGlobeTarget.clone().normalize());
+
```
Same : add comment please

gchoqueux replied:
This part must be reworked

peppsac commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -471,41 +469,39 @@ function GlobeControls(camera, domElement, engine) {
             // angle from z-axis around y-axis
-            if(sphericalDelta.theta || sphericalDelta.phi )
+            if(sphericalDelta.theta || sphericalDelta.phi ) {
+
```
Remove blank lines

peppsac commented on src/Renderer/ThreeExtented/GlobeControls.js:
```diff
@@ -1179,9 +1144,13 @@ GlobeControls.prototype.setHeading = function(heading) {
     this.updateCameraTransformation();
 };
 
+GlobeControls.prototype.getCenter = function() {
```
`getCenter`: center of what? center of view? on the globe?
Add a single line of comment or rename using a more descriptive  name

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -137,9 +137,16 @@ ApiGlobe.prototype.moveLayerToIndex = function(layer, newId) {
 ApiGlobe.prototype.removeImageryLayer = function(id) {
 
-    if (this.scene.getMap().removeColorLayer(id)) {
+    if (this.scene.getMap().layersConfiguration.removeColorLayer(id)) {
```
This doesn't fix half of the bugs in `removeImageryLayer` sadly - but I can wait for another PR to fix this.

gchoqueux replied:
Yes in is fixed in #200

peppsac replied:
Ok 👍

gchoqueux commented:
updated

gchoqueux committed:
Add ApiGlobe.setZoomLevel()
Add ApiGlobe.getZoomScale()
Add ApiGlobe.setZoomScale(zoomScale)
Add ApiGlobe.pan(pVector)

review

gchoqueux committed:
Remove small oblique weight in Screen space error compute
Oblique weight is wrong for tiles with big altitude

review sse

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -122,6 +121,12 @@ ApiGlobe.prototype.moveLayerDown = function(layer) {
 };
 
+/**
+ * { Moves a specific layer to a specific index in the layer list. This function has no effect if the layer is moved to its current index. }
```
Could you remove the `{ }` ?

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -122,6 +121,12 @@ ApiGlobe.prototype.moveLayerDown = function(layer) {
+ * { Moves a specific layer to a specific index in the layer list. This function has no effect if the layer is moved to its current index. }
+ *
+ * @param      {string}  layer   The layer
+ * @param      {number}  newId   The new identifier
```
Incorrect parameter names:
- `layer` →`layerId`
- `newId` →`newIndex` (and change the documentation accordingly)

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -170,39 +182,34 @@ ApiGlobe.prototype.addElevationLayer = function(layer) {
+    id = id || 0;
+    var layer = this.getLayers()[id];
+    if(layer)
+        return layer.zoom.min;
```
There are several problems with this function.
1) what is `id` ? `layerIndex` would be a better name (as far as I can tell)
2) what if `id` is invalid ?
3) what if this layer has no zoom property (a `wms` layer for instance ?)
4) no need to modify id, you could use `this.getLayers()[id || 0]`
5) given the function name, maybe you could return the min zoom value for all layers with a zoom value if id is undefined?

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -170,39 +182,34 @@ ApiGlobe.prototype.addElevationLayer = function(layer) {
+    id = id || 0;
+    var layer = this.getLayers()[id];
+    if(layer)
+        return layer.zoom.max;
```
See comments for getMinZoomLevel

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -170,39 +182,34 @@ ApiGlobe.prototype.addElevationLayer = function(layer) {
 
-ApiGlobe.prototype.getMaxZoomLevel = function(id) {
-    //console.log(this.addImageryLayer().id);
+ApiGlobe.prototype.getLayers = function( /*param*/ ) {
```
`getColorLayers` ? `getImageryLayers` ?

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -170,39 +182,34 @@ ApiGlobe.prototype.addElevationLayer = function(layer) {
- * <iframe width="100%" height="400" src="//jsfiddle.net/iTownsIGN/y1xcqv4s/embedded/" allowfullscreen="allowfullscreen" frameborder="0"></iframe>
+ * Return the list of all layers in the scene in the order of how they are stacked on top of each other.
  * @constructor
  * @param {id} id - The id of the layer.
```
Remove invalid comments

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -456,16 +463,17 @@ ApiGlobe.prototype.getHeading = function() {
+    return this.scene.currentControls().getRange();
+};
 
+ApiGlobe.prototype.getRangeFromEllispoid = function() {
```
You didn't correct the typo `getRangeFromEllipsoid`

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
-ApiGlobe.prototype.setRange = function(pRange /*, bool*/ ) {
+ApiGlobe.prototype.setRange = function(pRange /*, bool anim*/) {
     eventRange.oldRange = this.getRange();
+    loaded = false;
```
Are you sure this is a good idea? I assumed that the `globe-loaded` event was supposed to be fired just once.
If that's not the case maybe documenting each event would be a good idea.

gchoqueux replied:
Yes, the event represents the tiles are loaded. In this case, I need to know when the tiles are loaded to compute the new target camera.

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
+};
+
+/**
+ * { Displaces the central point to a specific amount of pixels from its current position.
```
Remove `{  }`

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
+    let range = center.distanceTo(camera.camera3D.position);
+    // compute distance camera/globe's center
+    let distance = rayon + range;
+    // C is point in position camera
```
C = camera's position

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
+    // compute distance camera/globe's center
+    let distance = rayon + range;
+    // C is point in position camera
+    // G is point in globe's center
```
G = globe's center

peppsac commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -565,28 +573,150 @@ ApiGlobe.prototype.setCenterAdvanced = function(pPosition /*, pDisableAnimationo
+    let distance = rayon + range;
+    // C is point in position camera
+    // G is point in globe's center
+    // S is point intersection with surface globe and camera (the furthest from line CG)
```
S = the furthest interesection[camera frustrum, globe surface]  from line CG

peppsac commented on src/Renderer/Camera.js:
```diff
@@ -116,19 +116,20 @@ Camera.prototype.resize = function(width, height) {
     var SSE = Math.sqrt(dotProductW) * this.preSSE * (node.geometricError / distance);
-    //var SSE = this.preSSE * (node.geometricError / distance);
+    */
+    var SSE = this.preSSE * (node.geometricError / distance);
```
Could you add a comment here:

```
// TODO: node.geometricError is computed using a hardcoded 18 level
```

peppsac commented on src/Scene/Quadtree.js:
```diff
@@ -28,8 +29,7 @@ function commandQueuePriorityFunction(cmd) {
         } else {
-            // TODO: this magic value comes from NodeProcess
-            if (6.0 < node.sse) {
+            if (SSE_SUBDIVISION_THRESHOLD < node.sse) {
```
👍

peppsac commented:
Looking better, thanks. Merging
